### PR TITLE
ERE-1876] Add form position saving

### PR DESCRIPTION
**Pull request by ScriptSmith**
TODO: 

- [x] Determine sections displayed in viewport using `IntersectionObserver`
- [x] Include form position in POST form data
- [x] Scroll to form position on load
- [x] Wait for all bootstrap collapse events before setting position
- [x] When multiple sections have 100% their entire body visible in the viewport, pick the highest one.
- [x] Make position saving optional

### CODE REVIEW

**Comment by ScriptSmith** [comment]
Still needs some final touches & testing, but I'm interested in feedback on the feature.

I ended up needing to use [InteractionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) rather than traditional APIs because it was proving too difficult to keep track of focus / interaction with the form elements any other way. I hope the support and performance of this is OK, I haven't tried it out on a phone / low-spec device yet.

Likewise I also needed to use [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) because [waiting for the collapse animations to complete](https://github.com/eresearchqut/trrf/pull/487/commits/e111822543eceb4e89b5e2b367fcba0d3e1e5f0a) wasn't enough to initiate the scroll, as the CDE widgets mess with page height as they're being rendered so it would end up scrolling to the wrong place.

Also, the banner being `position: fixed` makes scrolling to a particular element a bit challenging, as if you use the default `scrollIntoView` config, it will put scroll the page so that element is obscured by the banner. Scrolling so that the element is centered seems OK, although it can be a little disorientating. The fixed banner also means that `ResizeObserver` might think an element is 100% visible when it isn't. 

I've made the feature optional, because the default behaviour is to collapse all sections on save (unless there are errors). We could consider changing this, but I think it's something angelman needed for quick navigation, so I'm apprehensive about that.

Overall it feels a bit hacky & tacked-on. To do this properly, I think you'd need to submit the form data using JS rather than as a form `POST`, but that would be a much larger change.

**Comment by ScriptSmith** [comment]
Updated to just focus on interactions with form input elements as suggested by @ppettitau 

Now captures the current position on redirect when a form is saved for the first time.

Also scrolls to a visible ancestor when there are hidden form elements that have events triggered.

As discussed, we could make this the default behaviour, but this has only been requested because of UX in a clinical setting with a couple of large forms, so I think making it optional is best for now.

File: rdrf/rdrf/models/definition/models.py [hunk 1]
```diff
@@ -831,6 +831,7 @@ class RegistryForm(models.Model):
                      Click <a href="/forms/dsl-help" target="_blank">here</a> for more info'''
     )
     tags = ArrayField(models.CharField(max_length=20), default=list, blank=True)
+    save_position = models.BooleanField(default=False, help_text="Return to the current position on form save")
```
Comment (ppettitau) [inline rdrf/rdrf/models/definition/models.py]: Would this benefit from being a bit more verbose. e.g. "Return the user to their current position on form save"?

File: rdrf/rdrf/templates/rdrf_cdes/form.html [hunk 2]
```diff
@@ -113,6 +113,19 @@
...
+        rdrfSetupFormPositionSaving();
+
+        {% if current_position %}
+            rdrfScrollToFormPosition("{{ current_position }}");
+        {% else %}
+            var searchParams = new URLSearchParams(window.location.search);
+            var currentPosition = searchParams.get("currentPosition")
+            if (currentPosition) {
+                searchParams.delete("currentPosition");
+                window.history.replaceState(null, null, "?" + searchParams);
```
Comment (ppettitau) [inline rdrf/rdrf/templates/rdrf_cdes/form.html:125]: Cool, didn't know you could do this.

**Review by ppettitau — CHANGES_REQUESTED** [review]
Looks like a good solution to a complex problem.

Just had the one question about the save_position label.

File: rdrf/rdrf/templates/rdrf_cdes/form.html (hunk 2 above)
Comment (ScriptSmith) [reply rdrf/rdrf/templates/rdrf_cdes/form.html:125]: Neither did I :)

File: rdrf/rdrf/models/definition/models.py (hunk 1 above)
Comment (ScriptSmith) [reply rdrf/rdrf/models/definition/models.py]: Agreed

**Review by ppettitau — APPROVED** [review]

**Pull request merged**